zarr/core.py

```diff
@@ -627,125 +751,128 @@ def _setitem_zd(self, item, value):
 
-                # put data
-                self._chunk_setitem(cidx, chunk_selection, value)
+        # iterate over chunks in range
+        for chunk_coords, chunk_selection, out_selection in indexer:
 
+            # extract data to store
+            if np.isscalar(value):
+                chunk_value = value
             else:
-                # assume value is array-like
-
-                # determine index within value
-                value_selection = tuple(
-                    slice(max(0, o - s.start),
-                          min(o + c - s.start, s.stop - s.start))
-                    for s, o, c in zip(selection, offset, self._chunks)
-                    if isinstance(s, slice)
-                )
-
-                # put data
-                self._chunk_setitem(cidx, chunk_selection, value[value_selection])
-
-    def _chunk_getitem(self, cidx, item, dest):
+                chunk_value = value[out_selection]
+                # handle missing singleton dimensions
+                if indexer.drop_axes:
+                    item = [slice(None)] * self.ndim
+                    for a in indexer.drop_axes:
+                        item[a] = np.newaxis
+                    chunk_value = chunk_value[item]
+
+            # put data
+            self._chunk_setitem(chunk_coords, chunk_selection, chunk_value)
+
+    def _chunk_getitem(self, chunk_coords, chunk_selection, out, out_selection, drop_axes=None):
         """Obtain part or whole of a chunk.
 
         Parameters
         ----------
-        cidx : tuple of ints
+        chunk_coords : tuple of ints
             Indices of the chunk.
-        item : tuple of slices
-            Location of region within the chunk.
-        dest : ndarray
-            Numpy array to store result in.
+        chunk_selection : selection
+            Location of region within the chunk to extract.
+        out : ndarray
+            Array to store result in.
+        out_selection : selection
+            Location of region within output array to store results in.
+        drop_axes : tuple of ints
+            Axes to squeeze out of the chunk.
 
         """
 
+        assert len(chunk_coords) == len(self._cdata_shape)
+
         try:
 
             # obtain compressed data for chunk
-            ckey = self._chunk_key(cidx)
+            ckey = self._chunk_key(chunk_coords)
             cdata = self.chunk_store[ckey]
 
         except KeyError:
 
             # chunk not initialized
             if self._fill_value is not None:
-                dest.fill(self._fill_value)
+                out[out_selection] = self._fill_value
 
         else:
 
-            if is_total_slice(item, self._chunks) and \
-                    not self._filters and \
-                    ((self._order == 'C' and dest.flags.c_contiguous) or
-                     (self._order == 'F' and dest.flags.f_contiguous)):
+            if isinstance(out, np.ndarray) and \
```

Comment:
note: PEP8 suggests using extra parentheses rather than explicit `\` for line continuation. I think it looks a little cleaner, too.